src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs

```diff
@@ -11,21 +12,91 @@ internal class RedisStreamTriggerScaleMonitor : RedisPollingTriggerBaseScaleMoni
+            if (streamLength == 0)
             {
-                Remaining = multiplexer.GetDatabase().StreamLength(key),
+                return new RedisPollingTriggerBaseMetrics
+                {
+                    Remaining = 0,
+                    Timestamp = DateTime.UtcNow,
+                };
+            }
+
+            StreamGroupInfo[] groups = multiplexer.GetDatabase().StreamGroupInfo(key);
+            if (!groups.Any(g => g.Name == name))
+            {
+                // group doesn't exist, assume all entries in stream have not been processed
+                return new RedisPollingTriggerBaseMetrics
+                {
+                    Remaining = streamLength,
+                    Timestamp = DateTime.UtcNow,
+                };
+            }
+
+            StreamGroupInfo group = groups.First(g => g.Name == name);
+            if (multiplexer.GetServers()[0].Version >= RedisUtilities.Version70 && group.Lag.HasValue)
+            {
+                // Redis 7: Scaler gets number of remaining entries for the consumer group from XINFO GROUPS.
+                return new RedisPollingTriggerBaseMetrics
+                {
+                    Remaining = group.Lag.Value,
+                    Timestamp = DateTime.UtcNow,
+                };
+            }
+
+            StreamInfo info = await multiplexer.GetDatabase().StreamInfoAsync(key);
+            string firstId = (string)info.FirstEntry.Id;
+            string lastId = (string)info.LastEntry.Id;
+            string lastDeliveredId = group.LastDeliveredId ?? firstId;
+            if (lastId == lastDeliveredId)
+            {
+                // processed all entries in the stream
+                return new RedisPollingTriggerBaseMetrics
+                {
+                    Remaining = 0,
+                    Timestamp = DateTime.UtcNow,
+                };
+            }
+
+            // Redis 6/6.2 does not have an internal counter for remaining entries for the consumer group
+            // Estimate remaining entries using the time field from the entry ID (assuming ID="time-counter")
+            string[] firstIdSplit = firstId.Split('-');
+            string[] lastIdSplit = lastId.Split('-');
+            string[] lastDeliveredIdSplit = lastDeliveredId.Split('-');
+            ulong firstTimestamp = ulong.Parse(firstIdSplit[0]);
+            ulong lastTimestamp = ulong.Parse(lastIdSplit[0]);
+            ulong lastDeliveredTimestamp = ulong.Parse(lastDeliveredIdSplit[0]);
+
+            if (lastTimestamp == lastDeliveredTimestamp)
+            {
+                // If timestamp is the same, return counter difference
+                ulong lastCounter = ulong.Parse(lastIdSplit[1]);
+                ulong lastDelieveredCoutner = ulong.Parse(lastDeliveredIdSplit[1]);
+                return new RedisPollingTriggerBaseMetrics
+                {
+                    Remaining = Math.Min(streamLength, (long)Math.Max(1, lastCounter - lastDelieveredCoutner)),
```

Comment:
There is precision loss from ulong to long, though users might unlikely run into it. Given the count of element in stream is `ulong` as defined in Redis, we shall definitely change our library across. I am fine we do it in a separate PR. It is your call

```c
typedef struct stream {
    rax *rax;               /* The radix tree holding the stream. */
    uint64_t length;        /* Current number of elements inside this stream. */
    streamID last_id;       /* Zero if there are yet no items. */
    streamID first_id;      /* The first non-tombstone entry, zero if empty. */
    streamID max_deleted_entry_id;  /* The maximal ID that was deleted. */
    uint64_t entries_added; /* All time count of elements added. */
    rax *cgroups;           /* Consumer groups dictionary: name -> streamCG */
} stream;
```

Reply:
I will have to update this for the list length as well [since that is `ulong` as well](https://github.com/redis/redis/blob/38f02349462d5aefa9b25386a130ebd67db1a4de/src/adlist.h#L47), so I will combine both `long` to `ulong` conversions in a separate PR

Reply:
StackExchange.Redis uses `long` for both [ListLength](https://github.com/StackExchange/StackExchange.Redis/blob/b2694b385c876793a827bd4d2556439983061877/src/StackExchange.Redis/Interfaces/IDatabase.cs#L1027) and [StreamLength](https://github.com/StackExchange/StackExchange.Redis/blob/b2694b385c876793a827bd4d2556439983061877/src/StackExchange.Redis/Interfaces/IDatabase.cs#L2423C33-L2423C33), so I think it would be best to stay consistent with the SDK and leave it as is

Reply:
Thanks. @marcgravell @philon-msft , could you please provide any context here?